rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2150 @@
+
+#### `PromiseTransmutableFrom` and `PromiseTransmutableInto`
+
+You may declare the stability guarantees of your type by implementing one or both of two traits:
+```rust
+pub trait PromiseTransmutableFrom
+{
+    type Archetype
+        : TransmuteInto<Self, NeglectStability>
+        + PromiseTransmutableFrom;
+}
+
+pub trait PromiseTransmutableInto
+{
+    type Archetype
+        : TransmuteFrom<Self, NeglectStability>
+        + PromiseTransmutableInto;
+}
+```
+
+To implement each of these traits, you must specify an `Archetype`. An `Archetype` is a type whose layout exemplifies the extremities of your stability promise (i.e., the least/most constrained type for which it is valid to transmute your type into/from).
+
+By implementing `PromiseTransmutableFrom`, you promise that your type is guaranteed to be safely transmutable *from* `PromiseTransmutableFrom::Archetype`. Conversely, by implementing `PromiseTransmutableInto`, you promise that your type is guaranteed to be safely transmutable *into* `PromiseTransmutableInto::Archetype`.
+
+You are free to change the layout of your type however you like between minor crate versions so long as that change does not violates these promises. These two traits are capable of expressing simple and complex stability guarantees.
+
+#### Stability & Transmutation
+A `Src` type is *stably* transmutable into a `Dst` type *only if* `<Src as PromiseTransmutableInto>::Archetype` is transmutable, stability notwithstanding, into `<Dst as PromiseTransmutableFrom>::Archetype`; formally:
+```rust
+unsafe impl<Src, Dst> TransmuteFrom<Src> for Dst
+where
+    Src: PromiseTransmutableInto,
+    Dst: PromiseTransmutableFrom,
+    <Dst as PromiseTransmutableFrom>::Archetype:
+        TransmuteFrom<
+            <Src as PromiseTransmutableInto>::Archetype,
+            NeglectStability
+        >
+{}
+```
+
+Why is this the case?
+
+The type `<Src as PromiseTransmutableInto>::Archetype` exemplifies the furthest extreme of non-breaking changes that could be made to the layout of `Src` that could affect its use as a source type in transmutations. Conversely, `<Dst as PromiseTransmutableFrom>::Archetype` exemplifies the furthest extreme of non-breaking changes that could be made to the layout of `Dst` that could affect its use as a destination type in transmutations. If a transmutation between these extremities is valid, then so is `Src: TransmuteInto<Dst>`.
+
+#### Common Use-Case: Promising a Fixed Layout
+To promise that type's layout is completely fixed, simply `derive` the `PromiseTransmutableFrom` and `PromiseTransmutableInto` traits:
+```rust
+#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]
+#[repr(C)]
+pub struct Foo(pub Bar, pub Baz);
+```
+This expands to:
+```rust
+#[repr(C)]
+pub struct Foo(pub Bar, pub Baz);
+
+// Generated stability declarations for `Foo`:
+const _: () = {
```

Comment:
How can 2 custom derives result in only 1 top-level item?

Reply:
Whoops, you're right. This *should* result in two separate items.

Reply:
Fixed by https://github.com/rust-lang/project-safe-transmute/pull/5/commits/7a883702a5390f8419689dfc188f263a665ac871